From 0fa27e82724ac883d7b901203724f19729c0078d Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Wed, 29 Oct 2014 16:34:10 -0400
Subject: [fs] ext4: fix wrong size computation in ext4_mb_normalize_request()

Message-id: <1414600485-1929-21-git-send-email-lczerner@redhat.com>
Patchwork-id: 98537
O-Subject: [RHEL 7.1 PATCH v2 20/55] ext4: fix wrong size computation in ext4_mb_normalize_request()
Bugzilla: 1156096
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Xiaoguang Wang <wangxg.fnst@cn.fujitsu.com>

BZ 1156096
https://bugzilla.redhat.com/show_bug.cgi?id=1156096

BREW 8169722
https://brewweb.devel.redhat.com/taskinfo?taskID=8169722

Upstream commit b27b1535acc0e97b7435bccf205b3fe9c782afac

 As the member fe_len defined in struct ext4_free_extent is expressed as
 number of clusters, the variable "size" computation is wrong, we need to
 first translate fe_len to block number, then to bytes.

 Signed-off-by: Xiaoguang Wang <wangxg.fnst@cn.fujitsu.com>
 Signed-off-by: Theodore Ts'o <tytso@mit.edu>
 Reviewed-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/mballoc.c b/fs/ext4/mballoc.c
index bd78e92..0307faf 100644
--- a/fs/ext4/mballoc.c
+++ b/fs/ext4/mballoc.c
@@ -3073,8 +3073,9 @@ ext4_mb_normalize_request(struct ext4_allocation_context *ac,
        (23 - bsbits)) << 23;
   size = 8 * 1024 * 1024;
  } else {
-  start_off = (loff_t)ac->ac_o_ex.fe_logical << bsbits;
-  size   = ac->ac_o_ex.fe_len << bsbits;
+  start_off = (loff_t) ac->ac_o_ex.fe_logical << bsbits;
+  size   = (loff_t) EXT4_C2B(EXT4_SB(ac->ac_sb),
+           ac->ac_o_ex.fe_len) << bsbits;
  }
  size = size >> bsbits;
  start = start_off >> bsbits;
-- 
1.7.1